### PR TITLE
impl(oauth2): retrieve project id from MDS

### PR DESCRIPTION
**Pull request by coryan**
Part of the work for #14112

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14150)

### CODE REVIEW

**Review by devbww — APPROVED** [review]

File: google/cloud/internal/oauth2_compute_engine_credentials.cc [hunk 1]
```diff
@@ -225,6 +225,16 @@ StatusOr<std::string> ComputeEngineCredentials::universe_domain(
   return RetrieveUniverseDomain(lock, options);
 }
 
+StatusOr<std::string> ComputeEngineCredentials::project_id() const {
+  return project_id(Options{});
+}
+
+StatusOr<std::string> ComputeEngineCredentials::project_id(
+    google::cloud::Options const& options) const {
+  std::lock_guard<std::mutex> lk(mu_);
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/oauth2_compute_engine_credentials.cc:234]: As in maybe we should have a mutex for `project_id_` and one for `universe_domain_` and so forth?  It probably won't matter (these things are called rarely), but it is not too terrible to change this.

[the remark this answers is not in the document]
Comment (devbww) [reply google/cloud/internal/oauth2_compute_engine_credentials.cc:234]: Yes, that was the drift.  But I'm happy with _"it probably won't matter."_  Thanks.

[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/oauth2_compute_engine_credentials.cc]: Done. While it is unlikely to be a problem, I would rather not answer the pager when somebody reports contention here.

[the remark this answers is not in the document]
Comment (devbww) [reply google/cloud/internal/oauth2_compute_engine_credentials.cc]: Cool.

**Pull request merged**